OpenRA.Mods.D2k/UtilityCommands/D2kMapImporter.cs

```diff
@@ -0,0 +1,415 @@
+
+			terrainTiles = new List<TerrainTile>();
+			specialValueForTileIndex = new Dictionary<int, int>();
+
+			// Get all templates from the tileset YAML file that have at least one frame and an Image property corresponding to the requested tileset 
+			// Each frame is a tile from the Dune 2000 tileset files, with the Frame ID being the index of the tile in the original file
+			tileSetsFromYaml = tileSet.Templates.Where(t => t.Value.Frames != null
+				&& t.Value.Image.ToLower() == tilesetName.ToLower()).Select(ts => ts.Value).ToList();
+		}
+
+		void FillMap()
+		{
+			for (var y = 0; y < map.MapSize.Y; y++)
+				for (var x = 0; x < map.MapSize.X; x++)
+				{
+					var locationOnMap = new CPos(x, y);
+
+					// Hack for a "clear" tile
+					var indexOnMap = 0;
+
+					if (map.Contains(locationOnMap))
+						indexOnMap = (y - MapCordonWidth) * mapSize.Width + (x - MapCordonWidth); // +1;
+
+					if (indexOnMap < 0 || indexOnMap >= terrainTiles.Count)
+						indexOnMap = 0;
+
+					map.MapTiles.Value[locationOnMap] = terrainTiles[indexOnMap];
+					
+					// Tile special values
+					if (!specialValueForTileIndex.ContainsKey(indexOnMap))
+						continue;
+
+					var tileSpecialInfo = specialValueForTileIndex[indexOnMap];
+
+					// Spice
+					if (tileSpecialInfo == 1)
+						map.MapResources.Value[locationOnMap] = new ResourceTile(1, 1);
+					if (tileSpecialInfo == 2)
+						map.MapResources.Value[locationOnMap] = new ResourceTile(1, 2);
+
+					// Actors
+					if (actorDataByActorCode.ContainsKey(tileSpecialInfo))
+					{
+						var kvp = actorDataByActorCode[tileSpecialInfo];
+						if (!rules.Actors.ContainsKey(kvp.First.ToLower()))
+							throw new InvalidOperationException("Actor with name {0} could not be found in the rules YAML file!".F(kvp.First));
+
+						var a = new ActorReference(kvp.First)
+						{
+							new LocationInit(locationOnMap),
+							new OwnerInit(kvp.Second)
+						};
+						map.Actors.Value.Add("Actor" + map.Actors.Value.Count, a);
+					}
+				}
+		}
+
+		private TerrainTile GetTile(int tileIndex)
+		{
+			// Get the first tileset template that contains the Frame ID of the original map's tile with the requested index
+			var template = tileSetsFromYaml.FirstOrDefault(x => x.Frames.Contains(tileIndex));
+
+			// HACK: The arrakis.yaml tileset file seems to be missing some tiles, so just get a replacement for them
+			// Also used for duplicate tiles that are taken from only tileset
+			if (template == null)
+			{
+				// Just get a template that contains a tile with the same ID as requested
+				var templates = tileSet.Templates.Where(t => t.Value.Frames != null && t.Value.Frames.Contains(tileIndex));
+				if (templates.Any())
+					template = templates.First().Value;
+			}
+
+			if (template == null)
+				throw new InvalidOperationException("Tile with index {0} could not be found in the tileset YAML file!".F(tileIndex));
+
+			var templateIndex = template.Id;
+			byte frameIndex = 0;
```

Comment:
Yes, though if I understand the rest of the code correctly, you've already assured that tileIndex is in the array anyway.